File: src/apps/daemon/internal/networking/transport/quic_packet_codec.go
package transport

import (
	"encoding/binary"
	"errors"
)

type QuicHeaderType int

const (
	QuicPktInitial QuicHeaderType = iota
	QuicPktZeroRtt
	QuicPktHandshake
	QuicPktRetry
	QuicPkt1RttShort
)

type QuicPacketHeader struct {
	HeaderType   QuicHeaderType `json:"header_type"`
	Version      uint32         `json:"version"`
	DestCID      []byte         `json:"dest_cid"`
	SrcCID       []byte         `json:"src_cid"`
	PacketNumber uint64         `json:"packet_number"`
}

type QuicPacketCodec struct{}

func (q *QuicPacketCodec) EncodeVarint(value uint64) []byte {
	if value < (1 << 6) {
		return []byte{byte(value)}
	} else if value < (1 << 14) {
		return []byte{0x40 | byte(value>>8), byte(value & 0xff)}
	} else if value < (1 << 30) {
		return []byte{
			0x80 | byte(value>>24),
			byte((value >> 16) & 0xff),
			byte((value >> 8) & 0xff),
			byte(value & 0xff),
		}
	} else {
		out := []byte{0xc0 | byte(value>>56)}
		for i := 6; i >= 0; i-- {
			out = append(out, byte((value>>(i*8))&0xff))
		}
		return out
	}
}

func (q *QuicPacketCodec) DecodeVarint(data []byte) (uint64, int, error) {
	if len(data) == 0 {
		return 0, 0, errors.New("empty varint input")
	}

	prefix := data[0] >> 6
	switch prefix {
	case 0:
		return uint64(data[0]), 1, nil
	case 1:
		if len(data) < 2 {
			return 0, 0, errors.New("varint truncated at 2 bytes")
		}
		val := (uint64(data[0]&0x3f) << 8) | uint64(data[1])
		return val, 2, nil
	case 2:
		if len(data) < 4 {
			return 0, 0, errors.New("varint truncated at 4 bytes")
		}
		val := (uint64(data[0]&0x3f) << 24) |
			(uint64(data[1]) << 16) |
			(uint64(data[2]) << 8) |
			uint64(data[3])
		return val, 4, nil
	case 3:
		if len(data) < 8 {
			return 0, 0, errors.New("varint truncated at 8 bytes")
		}
		val := uint64(data[0]&0x3f) << 56
		for i := 1; i < 8; i++ {
			val |= uint64(data[i]) << ((7 - i) * 8)
		}
		return val, 8, nil
	}
	return 0, 0, errors.New("invalid varint prefix")
}

func (q *QuicPacketCodec) EncodePacket(header QuicPacketHeader, payload []byte) []byte {
	var out []byte

	if header.HeaderType == QuicPkt1RttShort {
		// Short header: Form bit 0, Fixed bit 1
		firstByte := byte(0x40 | 0x01)
		out = append(out, firstByte)
		out = append(out, header.DestCID...)
		pnBuf := make([]byte, 2)
		binary.BigEndian.PutUint16(pnBuf, uint16(header.PacketNumber))
		out = append(out, pnBuf...)
		out = append(out, payload...)
		return out
	}

	// Long header
	typeBits := byte(0)
	switch header.HeaderType {
	case QuicPktInitial:
		typeBits = 0x00
	case QuicPktZeroRtt:
		typeBits = 0x10
	case QuicPktHandshake:
		typeBits = 0x20
	case QuicPktRetry:
		typeBits = 0x30
	}

	firstByte := byte(0x80 | 0x40 | typeBits)
	out = append(out, firstByte)

	verBuf := make([]byte, 4)
	binary.BigEndian.PutUint32(verBuf, header.Version)
	out = append(out, verBuf...)

	out = append(out, byte(len(header.DestCID)))
	out = append(out, header.DestCID...)

	out = append(out, byte(len(header.SrcCID)))
	out = append(out, header.SrcCID...)

	pnBuf := make([]byte, 2)
	binary.BigEndian.PutUint16(pnBuf, uint16(header.PacketNumber))

	lenVarint := q.EncodeVarint(uint64(len(pnBuf) + len(payload)))
	out = append(out, lenVarint...)

	out = append(out, pnBuf...)
	out = append(out, payload...)
	return out
}

func (q *QuicPacketCodec) DecodePacket(data []byte, destCidLen int) (QuicPacketHeader, []byte, error) {
	if len(data) == 0 {
		return QuicPacketHeader{}, nil, errors.New("empty packet")
	}

	firstByte := data[0]
	isLong := (firstByte & 0x80) != 0

	if isLong {
		if len(data) < 7 {
			return QuicPacketHeader{}, nil, errors.New("long header truncated")
		}

		typeBits := (firstByte & 0x30) >> 4
		hType := QuicPktInitial
		switch typeBits {
		case 0x00:
			hType = QuicPktInitial
		case 0x01:
			hType = QuicPktZeroRtt
		case 0x02:
			hType = QuicPktHandshake
		case 0x03:
			hType = QuicPktRetry
		}

		version := binary.BigEndian.Uint32(data[1:5])
		offset := 5

		dcidLen := int(data[offset])
		offset++
		if offset+dcidLen > len(data) {
			return QuicPacketHeader{}, nil, errors.New("dcid truncated")
		}
		dcid := make([]byte, dcidLen)
		copy(dcid, data[offset:offset+dcidLen])
		offset += dcidLen

		if offset >= len(data) {
			return QuicPacketHeader{}, nil, errors.New("scid len truncated")
		}
		scidLen := int(data[offset])
		offset++
		if offset+scidLen > len(data) {
			return QuicPacketHeader{}, nil, errors.New("scid truncated")
		}
		scid := make([]byte, scidLen)
		copy(scid, data[offset:offset+scidLen])
		offset += scidLen

		payloadLen, vlen, err := q.DecodeVarint(data[offset:])
		if err != nil {
			return QuicPacketHeader{}, nil, err
		}
		offset += vlen

		if offset+2 > len(data) {
			return QuicPacketHeader{}, nil, errors.New("pn truncated")
		}
		pn := uint64(binary.BigEndian.Uint16(data[offset : offset+2]))
		offset += 2

		payloadSize := int(payloadLen) - 2
		if payloadSize < 0 || offset+payloadSize > len(data) {
			return QuicPacketHeader{}, nil, errors.New("payload truncated")
		}
		payload := data[offset : offset+payloadSize]

		return QuicPacketHeader{
			HeaderType:   hType,
			Version:      version,
			DestCID:      dcid,
			SrcCID:       scid,
			PacketNumber: pn,
		}, payload, nil
	}

	// Short header
	offset := 1
	if offset+destCidLen+2 > len(data) {
		return QuicPacketHeader{}, nil, errors.New("short header truncated")
	}

	dcid := make([]byte, destCidLen)
	copy(dcid, data[offset:offset+destCidLen])
	offset += destCidLen

	pn := uint64(binary.BigEndian.Uint16(data[offset : offset+2]))
	offset += 2

	payload := data[offset:]
	return QuicPacketHeader{
		HeaderType:   QuicPkt1RttShort,
		Version:      0,
		DestCID:      dcid,
		SrcCID:       nil,
		PacketNumber: pn,
	}, payload, nil
}

File: src/packages/lumicore/src/transport/quic_packet_codec.rs
//! # QUIC Packet Codec
//!
//! Bit-level QUIC packet parser and serializer implementing Long Header (Initial, Handshake, Retry)
//! and Short Header (1-RTT) framing, RFC 9000 variable-length integers, and CID management.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum QuicHeaderType {
    Initial = 0x00,
    ZeroRtt = 0x01,
    Handshake = 0x02,
    Retry = 0x03,
    OneRttShort = 0x04,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct QuicPacketHeader {
    pub header_type: QuicHeaderType,
    pub version: u32,
    pub dest_cid: Vec<u8>,
    pub src_cid: Vec<u8>,
    pub packet_number: u64,
}

pub struct QuicPacketCodec;

impl QuicPacketCodec {
    pub fn encode_varint(value: u64) -> Vec<u8> {
        if value < (1 << 6) {
            vec![value as u8]
        } else if value < (1 << 14) {
            vec![0x40 | ((value >> 8) as u8), (value & 0xff) as u8]
        } else if value < (1 << 30) {
            let mut out = vec![0x80 | ((value >> 24) as u8)];
            out.push(((value >> 16) & 0xff) as u8);
            out.push(((value >> 8) & 0xff) as u8);
            out.push((value & 0xff) as u8);
            out
        } else {
            let mut out = vec![0xc0 | ((value >> 56) as u8)];
            for i in (0..7).rev() {
                out.push(((value >> (i * 8)) & 0xff) as u8);
            }
            out
        }
    }

    pub fn decode_varint(data: &[u8]) -> Result<(u64, usize), String> {
        if data.is_empty() {
            return Err("Unexpected EOF reading varint".to_string());
        }

        let prefix = data[0] >> 6;
        match prefix {
            0 => Ok((data[0] as u64, 1)),
            1 => {
                if data.len() < 2 {
                    return Err("Varint truncated at 2 bytes".to_string());
                }
                let val = (((data[0] & 0x3f) as u64) << 8) | (data[1] as u64);
                Ok((val, 2))
            }
            2 => {
                if data.len() < 4 {
                    return Err("Varint truncated at 4 bytes".to_string());
                }
                let val = (((data[0] & 0x3f) as u64) << 24)
                    | ((data[1] as u64) << 16)
                    | ((data[2] as u64) << 8)
                    | (data[3] as u64);
                Ok((val, 4))
            }
            3 => {
                if data.len() < 8 {
                    return Err("Varint truncated at 8 bytes".to_string());
                }
                let mut val = ((data[0] & 0x3f) as u64) << 56;
                for i in 1..8 {
                    val |= (data[i] as u64) << ((7 - i) * 8);
                }
                Ok((val, 8))
            }
            _ => unreachable!(),
        }
    }

    pub fn encode_packet(header: &QuicPacketHeader, payload: &[u8]) -> Vec<u8> {
        let mut out = Vec::new();

        match header.header_type {
            QuicHeaderType::OneRttShort => {
                // Short header: Form bit (0), Fixed bit (1), Spin (0), Reserved (00), KeyPhase (0), PN length (2 bytes = 01)
                let first_byte = 0x40 | 0x01;
                out.push(first_byte);
                out.extend_from_slice(&header.dest_cid);
                out.extend_from_slice(&(header.packet_number as u16).to_be_bytes());
                out.extend_from_slice(payload);
            }
            long_type => {
                // Long header: Form bit (1), Fixed bit (1), Long packet type (2 bits)
                let type_bits = match long_type {
                    QuicHeaderType::Initial => 0x00,
                    QuicHeaderType::ZeroRtt => 0x10,
                    QuicHeaderType::Handshake => 0x20,
                    QuicHeaderType::Retry => 0x30,
                    QuicHeaderType::OneRttShort => unreachable!(),
                };
                let first_byte = 0x80 | 0x40 | type_bits;
                out.push(first_byte);
                out.extend_from_slice(&header.version.to_be_bytes());

                // Dest CID
                out.push(header.dest_cid.len() as u8);
                out.extend_from_slice(&header.dest_cid);

                // Src CID
                out.push(header.src_cid.len() as u8);
                out.extend_from_slice(&header.src_cid);

                // Length of token + payload + PN
                let pn_bytes = (header.packet_number as u16).to_be_bytes();
                let length_varint = Self::encode_varint((pn_bytes.len() + payload.len()) as u64);
                out.extend_from_slice(&length_varint);

                out.extend_from_slice(&pn_bytes);
                out.extend_from_slice(payload);
            }
        }

        out
    }

    pub fn decode_packet(data: &[u8], dest_cid_len: usize) -> Result<(QuicPacketHeader, Vec<u8>), String> {
        if data.is_empty() {
            return Err("Empty packet".to_string());
        }

        let first_byte = data[0];
        let is_long_header = (first_byte & 0x80) != 0;

        if is_long_header {
            if data.len() < 7 {
                return Err("Long header truncated".to_string());
            }

            let type_bits = (first_byte & 0x30) >> 4;
            let header_type = match type_bits {
                0x00 => QuicHeaderType::Initial,
                0x01 => QuicHeaderType::ZeroRtt,
                0x02 => QuicHeaderType::Handshake,
                0x03 => QuicHeaderType::Retry,
                _ => unreachable!(),
            };

            let version = u32::from_be_bytes([data[1], data[2], data[3], data[4]]);
            let mut offset = 5;

            let dcid_len = data[offset] as usize;
            offset += 1;
            if offset + dcid_len > data.len() {
                return Err("DCID truncated".to_string());
            }
            let dest_cid = data[offset..offset + dcid_len].to_vec();
            offset += dcid_len;

            if offset >= data.len() {
                return Err("SCID len truncated".to_string());
            }
            let scid_len = data[offset] as usize;
            offset += 1;
            if offset + scid_len > data.len() {
                return Err("SCID truncated".to_string());
            }
            let src_cid = data[offset..offset + scid_len].to_vec();
            offset += scid_len;

            let (payload_len, varint_len) = Self::decode_varint(&data[offset..])?;
            offset += varint_len;

            if offset + 2 > data.len() {
                return Err("Packet number truncated".to_string());
            }
            let packet_number = u16::from_be_bytes([data[offset], data[offset + 1]]) as u64;
            offset += 2;

            let payload_size = (payload_len as usize).saturating_sub(2);
            if offset + payload_size > data.len() {
                return Err("Payload truncated".to_string());
            }
            let payload = data[offset..offset + payload_size].to_vec();

            Ok((
                QuicPacketHeader {
                    header_type,
                    version,
                    dest_cid,
                    src_cid,
                    packet_number,
                },
                payload,
            ))
        } else {
            // Short header
            let mut offset = 1;
            if offset + dest_cid_len + 2 > data.len() {
                return Err("Short header truncated".to_string());
            }
            let dest_cid = data[offset..offset + dest_cid_len].to_vec();
            offset += dest_cid_len;

            let packet_number = u16::from_be_bytes([data[offset], data[offset + 1]]) as u64;
            offset += 2;

            let payload = data[offset..].to_vec();
            Ok((
                QuicPacketHeader {
                    header_type: QuicHeaderType::OneRttShort,
                    version: 0,
                    dest_cid,
                    src_cid: Vec::new(),
                    packet_number,
                },
                payload,
            ))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_varint_codec() {
        let values = [0, 42, 63, 64, 1500, 16383, 16384, 1073741823, 1073741824];
        for &v in &values {
            let encoded = QuicPacketCodec::encode_varint(v);
            let (decoded, len) = QuicPacketCodec::decode_varint(&encoded).unwrap();
            assert_eq!(v, decoded);
            assert_eq!(encoded.len(), len);
        }
    }

    #[test]
    fn test_long_header_initial_packet() {
        let header = QuicPacketHeader {
            header_type: QuicHeaderType::Initial,
            version: 1,
            dest_cid: vec![0x01, 0x02, 0x03, 0x04],
            src_cid: vec![0x05, 0x06, 0x07, 0x08],
            packet_number: 100,
        };

        let payload = b"ping_quic_frame_payload";
        let encoded = QuicPacketCodec::encode_packet(&header, payload);

        let (decoded_hdr, decoded_payload) = QuicPacketCodec::decode_packet(&encoded, 4).unwrap();
        assert_eq!(decoded_hdr, header);
        assert_eq!(decoded_payload, payload);
    }

    #[test]
    fn test_short_header_1rtt_packet() {
        let header = QuicPacketHeader {
            header_type: QuicHeaderType::OneRttShort,
            version: 0,
            dest_cid: vec![0xaa, 0xbb, 0xcc, 0xdd],
            src_cid: Vec::new(),
            packet_number: 55,
        };

        let payload = b"app_data_stream";
        let encoded = QuicPacketCodec::encode_packet(&header, payload);

        let (decoded_hdr, decoded_payload) = QuicPacketCodec::decode_packet(&encoded, 4).unwrap();
        assert_eq!(decoded_hdr.header_type, QuicHeaderType::OneRttShort);
        assert_eq!(decoded_hdr.dest_cid, header.dest_cid);
        assert_eq!(decoded_hdr.packet_number, 55);
        assert_eq!(decoded_payload, payload);
    }
}
